rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,704 @@
+- Make the `@`-pattern consistent with destructuring
+- A syntax to index into lists, e.g. `[ 1 2 3 ].0 == 1`
+- Use `,` to delimit elements of a list expression, like `[ 1, 2, 3 ]`
+- Do something about `?` meaning two different things depending on where it occurs (`{ x ? "" }:` vs `x ? ""`)
+- Better support for static analysis
+- [Syntax for hexadecimal numbers](https://github.com/NixOS/nix/pull/7695)
+
+Other discussions around language changes:
+
+- [Make path semantics less surprsing](https://github.com/NixOS/nix/issues/7338)
+- [RFC 110](https://github.com/NixOS/rfcs/pull/110) proposing a alternatives to the `with` expression
+- [Nix 2 – a hypothetical syntax](https://md.darmstadt.ccc.de/s/nix2)
+- [Nix language changes](https://gist.github.com/edolstra/29ce9d8ea399b703a7023073b0dbc00d)
+
+# Detailed Design
+
+## Design goals
+
+1. New versions of the Nix language evaluator should to stay backward compatible with existing Nix expressions.
+
+1. A Nix expression written in a newer version of the language should never work with older evaluators not supporting that version.
+
+1. Changes to the language, especially backward-incompatible changes, should remain a rare exception.
+
+## Language versioning
+
+1. The language version is a natural number.
+
+   <details><summary>Arguments</summary>
+
+   - (+) Formally decouples the Nix language version from the Nix version
+     - (+) The Nix language is supposed to change much less often than the rest of Nix
+     - (-) There are two version numbers to keep track of
+     - (-) Makes more evident that the Nix language is a distinct architectural component of the Nix ecosystem
+   - (+) It's currently handled that way, no change needed apart from documentation
+        - See [`builtins.langVersion`] (currently undocumented)
+   - (+) Simple and unambiguous
+   - (+) Concise, even in the long term, since the language is supposed to change very rarely
+
+   [`builtins.langVersion`]: https://github.com/NixOS/nix/blob/26c7602c390f8c511f326785b570918b2f468892/src/libexpr/primops.cc#L3952-L3957
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - [Calendar Versioning](https://calver.org/)
+       - (+) Provides information on when changes happened
+           - (-) This is not needed because only compatibility information is needed
+       - (-) Requires a minimum amount of characters
+           - This may be relevant depending on where it has to be encoded
+       - (+) Restricting to only the year would force language changes to be rare
+           - (+) This would allow obvious synchronisation points with Nixpkgs releases
+           - (-) It may be too much policy encoded in a mechanism
+   - [Semantic Versioning](https://semver.org/)
+       - (+) Can distinguish additions from other changes
+           - (-) This is not needed for our use case, since any addition to an expression will break for older evaluators even if the major version matches
```

Comment:
```suggestion
       - (+) Can distinguish additions from other changes
           - (-) This is not needed for our use case, since any addition to an expression will break for older evaluators even if the major version matches
       - (+) Simplifies evaluator implementation: pure additions can be guarded by the minor version cleanly within a major version's evaluator code
```